x/cron/keeper/keeper.go

```diff
@@ -0,0 +1,205 @@
+	}
+
+	return &Keeper{
+		cdc:           cdc,
+		storeKey:      storeKey,
+		memKey:        memKey,
+		paramstore:    ps,
+		accountKeeper: accountKeeper,
+	}
+}
+
+func (k *Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// ExecuteReadySchedules gets all schedules that are due for execution (with limit that is equals to Params.Limit)
+// and executes messages in each one
+// NOTE that errors in contract calls do NOT stop schedule execution
+func (k *Keeper) ExecuteReadySchedules(ctx sdk.Context) {
+	telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), LabelCheckTimer)
+	schedules := k.getSchedulesReadyForExecution(ctx)
+
+	for _, schedule := range schedules {
+		k.executeSchedule(ctx, schedule)
+	}
+}
+
+// AddSchedule adds new schedule to execution for every block `period`.
+// First schedule execution is supposed to be on `now + period` block.
+func (k *Keeper) AddSchedule(ctx sdk.Context, name string, period uint64, msgs []types.MsgExecuteContract) {
+	schedule := types.Schedule{
+		Name:              name,
+		Period:            period,
+		Msgs:              msgs,
+		LastExecuteHeight: uint64(ctx.BlockHeight()), // let's execute newly added schedule on `now + period` block
+	}
+	k.storeSchedule(ctx, schedule)
+}
+
+// RemoveSchedule removes schedule with a given `name`
+func (k *Keeper) RemoveSchedule(ctx sdk.Context, name string) {
+	k.removeSchedule(ctx, name)
+}
+
+// GetAllSchedules returns all schedules
+func (k *Keeper) GetAllSchedules(ctx sdk.Context) []types.Schedule {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+
+	res := make([]types.Schedule, 0)
+
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var schedule types.Schedule
+		k.cdc.MustUnmarshal(iterator.Value(), &schedule)
+		res = append(res, schedule)
+	}
+
+	return res
+}
+
+// GetSchedule returns schedule with a given `name`
+func (k *Keeper) GetSchedule(ctx sdk.Context, name string) (*types.Schedule, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+	bzSchedule := store.Get(types.GetScheduleKey(name))
+	if bzSchedule == nil {
+		return nil, false
+	}
+
+	var schedule types.Schedule
+	k.cdc.MustUnmarshal(bzSchedule, &schedule)
+	return &schedule, true
+}
+
+func (k *Keeper) getSchedulesReadyForExecution(ctx sdk.Context) []types.Schedule {
+	params := k.GetParams(ctx)
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+	count := uint64(0)
+
+	res := make([]types.Schedule, 0)
+
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var schedule types.Schedule
+		k.cdc.MustUnmarshal(iterator.Value(), &schedule)
+
+		if k.intervalPassed(ctx, schedule) {
+			res = append(res, schedule)
+			count++
+
+			if count >= params.Limit {
+				return res
```

Comment:
Should we warn about the limit reached. There is a possible corner case - 10 schedules with 3 block interval each. The last schedule will never be executed and its to easier to know about with some warnings.